From 1ba8f7df4e7720ac86c0e95ad47f52db3b176a7d Mon Sep 17 00:00:00 2001
From: Sadaf Ebrahimi <sadafebrahimi@google.com>
Date: Thu, 2 Jun 2022 19:32:22 +0000
Subject: [PATCH] Prevent XML_GetBuffer signed integer overflow

Bug: http://b/221255869
Change-Id: I38758fae8c71184f728f95e6073457cdb86bcc29
(cherry picked from commit a924d536de39ee21d4ab7051a3d13684162259bf)
Merged-In: I38758fae8c71184f728f95e6073457cdb86bcc29
---
 lib/xmlparse.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/xmlparse.c b/lib/xmlparse.c
index 02bcfaa..78ad787 100644
--- a/lib/xmlparse.c
+++ b/lib/xmlparse.c
@@ -1967,6 +1967,11 @@ XML_GetBuffer(XML_Parser parser, int len) {
     keep = (int)EXPAT_SAFE_PTR_DIFF(parser->m_bufferPtr, parser->m_buffer);
     if (keep > XML_CONTEXT_BYTES)
       keep = XML_CONTEXT_BYTES;
+    /* Detect and prevent integer overflow */
+    if (keep > INT_MAX - neededSize) {
+      parser->m_errorCode = XML_ERROR_NO_MEMORY;
+      return NULL;
+    }
     neededSize += keep;
 #endif /* defined XML_CONTEXT_BYTES */
     if (neededSize
-- 
2.37.1.359.gd136c6c3e2-goog

